WIP keymap functions

sehnsucht13 opened the pull request:
I have done some rough work on porting some keymap functions but I would love some feedback on the work so far. Right now, none of them have been tested but I am wondering if I am on the right track in terms of it being a faithful port(good rust practices) and any other general tips. I am quite new to rust so I am still trying to get comfortable with it.

There are some comments in the code which indicate what I would like to have looked at. One of the most troublesome functions is the `_copy_keymap_item` which is a mess of `unwrap` and `as_cons()`. I am hoping that there is a cleaner way or at the very least know that this is the most optimal way currently. Thanks in advance.

sehnsucht13 committed:
port map_keymap_call and map_keymap_canonicalize

- Still need to replace the function calls to these functions inside
other rust functions

sehnsucht13 committed:
Add skeleton for getkeyelt. Add is_not_cons

sehnsucht13 committed:
finish porting get_keyelt

sehnsucht13 committed:
port copy-keymap-item

sehnsucht13 committed:
Add is_not_integer. Port more keymap functions

- The function are not fully complete but remacs builds in its current
state.

sehnsucht13 committed:
Fix scoping in _get_keyelt. Add comments to _get_keyelt. Format

sehnsucht13 committed:
Add click_position, add getters for zv and pt in buffer

sehnsucht13 committed:
Format buffer

sehnsucht13 committed:
Remove unfinished functions for review

sehnsucht13 committed:
remove more dead comments

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -19,22 +20,24 @@ use crate::{
+    marker::marker_position,
     multibyte::LispStringRef,
     obarray::intern,
     remacs_sys::{
```
Things like `call2` and `list2` are almost always the wrong thing in the Rust code. There is usually a function or a macro to mimic the C code's needs.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -19,22 +20,24 @@ use crate::{
-    remacs_sys::{char_bits, current_global_map as _current_global_map, globals, EmacsInt},
+    remacs_sys::{char_bits, current_global_map as _current_global_map,
+    globals, EmacsInt, PURE_P},
     remacs_sys::{
```
Whenever you look to import an `F` function check if it has been re-implemented in Rust and call the Rust native function instead. `Flength` for instance. `Fcons` can usually be replicated by making a tuple and calling `into`.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+This can GC because menu_item_eval_property calls Feval.  */
+#[no_mangle]
+pub extern "C" fn _get_keyelt(object: LispObject, autoload: bool) -> LispObject {
+    loop {
```
```suggestion
    while let Some(cons) = object.as_cons() {
```

Does that read better? Then if the "if not cons" and "as_cons" below are not needed.

shaleh replied:
Actually, reading below, it appears we could go a little further and instead do:

`while let Some(car, cdr) = object.into() {`

However, I would suggest we find a better name for `car` and `cdr` here.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+        let mut localObject = object;
+
+        if consObject.car().eq(Qmenu_item) {
+            if consObject.cdr().is_cons() {
```
```suggestion
            if let Some(blah, blah) = consObject.cdr().into() {
```

Or use a match

```
match consObject.cdr().into() {
    Some(blah, blah) => ......
    None => break localObject;
}
```

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+#[no_mangle]
+pub extern "C" fn _copy_keymap_item(elt: LispObject) -> LispObject {
+    if elt.is_not_cons() {
+        elt;
```
? Are you trying to return `elt` here? If you the return keyword is needed.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+}
+
+// TODO: Fix this mess with all the unwrap but how?
+// TODO: Avoid using FCONS?
```
Read my guidance earlier and suggestions above. Should make this code more manageable.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+    } else {
+        curr_pos = ThreadState::current_buffer_unchecked()
+            .get_pt()
+            .try_into()
```
try_into without actually handling failures is something we would like to avoid. This can lead to Remacs crashing unexpectedly.

sehnsucht13 commented:
Thanks a lot for the review! I will work on making the changes!

shaleh commented:
@sehnsucht13 I avoided reviewing the actual port and commented on your mechanics. That seemed more helpful at this stage.

For C functions, if no C code calls them feel empowered to refactor them as needed.

sehnsucht13 commented:
Thanks again. I appreciate you taking the time to do this!

sehnsucht13 commented:
In the function `_copy_keymap_item`, there is a large and unsightly(in my opinion) usage of `obj.as_cons().unwrap()`. Should we add a function that is called `as_cons_unsafe()` which does this under the hood in order to make the code more readable or would that be a dangerous choice?

shaleh commented:
There is a definition for `From<LIspObject> for (LispObject, LispObject)`. Every Cons cell is really a tuple internally of (car, cdr). A Lisp list looks like `(car, (car, (car, (car, cdr))))`.

sehnsucht13 committed:
Remove try_into. Rewrite get_keyelt and copy_keymap_item

- Remove try into from from _click_position by casting to emacsint
- Use if let in get_keyelt and copy_keymap_item to perform checking of
- values and for extra readablity.
- Rewrite get_keyelt and copy_keymap_item to be shorter

sehnsucht13 committed:
Run rust format to pass check for circleci

sehnsucht13 commented:
I implemented the changes mentioned. 
- I have gotten rid of `try_into` and it turns out that a simple cast to EmacsInt will do the trick. I have no idea what I was thinking at the time since we are just comparing a ptrdiff_t with a usize.
- I have rewritten the two functions `get_keyelt` and `copy_keymap_item` to use the `while let` and `if let` for (hopefully) easier readability than before. The namings(head1,2...) are things which still need to be changed to something a bit more meaningful.

There is only one problem and that has to do with the `copy_keymap_item` function. The original C function uses 3 temporary variables called `tem`, `elt` and `res`. The `res` variable contains the value returned from the function but `elt` and `res` seem to be used to simply iterate over the cons cells. I have eliminated the use of `tem` since it was used to destructure each lisp object into a cons cell(I do that with if let). On the other hand, I am not too sure about elt itself and what its doing. Is it just a helper for `tem` or am I missing something here? Thanks in advance

sehnsucht13 commented:
Another question as well. What do we do about testing the rust code that has been ported? Some of it is accessible from emacs itself so `ert` can be used but what about functions which are not? Can we write unit tests for those?

sehnsucht13 committed:
Remove unused import

shaleh commented:
At the moment we cannot test Rust functions if they use ANY C types. Which basically means we cannot unit test.

I have this PR https://github.com/remacs/remacs/pull/1511 which needs someone's time. It makes it possible to run tests from C that are mostly implemented as Rust tests.

sehnsucht13 commented:
Build fails. I will take a look at it tommorow and fix it up

sehnsucht13 commented:
Just a bit of an update on this. I have not forgotten about it but I have school finals for summer semester coming up so this will have to wait for a bit.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +63,177 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+
+                        unsafe {
+                            filter = list!(filter, list!(Qquote, object));
+                            object = menu_item_eval_property(filter);
```
I think because `object` is redefined within the while look on line 110. The value returned on 155 is the original value.

shaleh commented on rust_src/src/keymap.rs:
```diff
@@ -62,6 +63,177 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+
+#[no_mangle]
+pub extern "C" fn _copy_keymap_item(elt: LispObject) -> LispObject {
+    if let Some((head, tail)) = elt.into() {
```
I would rewrite this for clarity.

a)

```
match elt.into() {
    None => { return elt; }
    Some((head, tail)) => {
        ...
    }
}
```
b)
```
let cons: Option<(LispObject, LispObject)> = elt.into();
if cons.is_none() {
    return elt;
}
let (head, tail) = cons;
.....
```
